README.md
<div style="text-align:center;" align="center">

# Node-Notifier

Open client that triggers desktop notifications through HTTP requests  
Intended to be used together with [ESP-Notifier](https://github.com/Sv443/ESP-Notifier) to send a desktop notification through an electrical button press

<br>

[![License](https://img.shields.io/github/license/Sv443/Node-Notifier)](https://sv443.net/LICENSE)
[![Known Vulnerabilities](https://snyk.io/test/github/Sv443/Node-Notifier/badge.svg)](https://snyk.io/test/github/Sv443/Node-Notifier)
[![Discord](https://img.shields.io/discord/565933531214118942)](https://dc.sv443.net/)

<br>

## Video Showcase:

[![video showcase (click to watch)](./.github/video_thumb.png)](https://youtu.be/)

</div>


<br>


## Table of Contents:
- **[Installation](#installation)**
    - [Notes](#notes)
- **[Other Info](#other-info)**
    - [Files and folders](#files-and-folders)
    - [Advanced configuration](#advanced-configuration)
- **[REST API usage](#rest-api-usage)**
    - [Sending a notification](#sending-a-notification)
- [Troubleshooting (help something broke)](#troubleshooting)
- [Disclaimers](#disclaimers)
- [Dependencies](#dependencies)



<br><br>


## Installation:
1. Install [Node.js](https://nodejs.org/)
2. Clone or download and extract this repo
3. Open a terminal in the project directory
4. Install dependencies with the command `npm i`
    - This also installs [pm2](https://npmjs.com/package/pm2) globally. It enables autostart on reboot, monitoring, etc.
5. Start Node-Notifier with `npm start`
    - To start it without being wrapped by pm2 (not recommended for normal use) you can use the command `npm run debug`
6. Enter new login info for the admin user
7. Now you're in the control panel, from here:
    - Test the notification system by selecting the option `Send test notification`
    - Open the dashboard by selecting the option `Open web dashboard` or opening [`http://127.0.0.1:8042`](http://127.0.0.1:8042/) in your browser
8. To make use of Node-Notifier, you can now do either of these:
    - Build an [ESP-Notifier](https://github.com/Sv443/ESP-Notifier) and configure it to work together with Node-Notifier
    - Send requests to the server in another way (see [REST API usage](#rest-api-usage))

<br>

I recommend you look at the rest of this document to familiarize yourself with Node-Notifier.

<br>

### Notes:
- To configure the internals of Node-Notifier even beyond what the dashboard and control panel can do, edit the file `config.js` in the root directory of Node-Notifier.  
After saving, if Node-Notifier is already running, make sure to restart its background process.  
If this is still not enough, you can follow the [advanced configuration section.](#advanced-configuration)
  
- If you need to change or delete your password, use the login manager in the control panel.  
If it's unavailable, you can use the command `npm run login-mgr` to open it instead.  
  
- If you need help with anything or just want to chat you can [join my Discord server.](https://dc.sv443.net/)


<br><br><br>


## Other Info:
Some miscellaneous info about Node-Notifier:

<br>

### Files and folders:
Node-Notifier generates a few files and folders:
- `assets/`  
    This folder is intended to be used as a place to store your notifications' images and potentially other assets.  
    It is generated and filled with an example icon when Node-Notifier first starts up.
  
- `.notifier/`  
    This folder contains all internal files of Node-Notifier.  
    Usually you shouldn't need to edit anything in here, but the program is open and I can't *really* stop you so go wild.
  
- `.notifier/properties.json`  
    This file contains persistent, internal data of Node-Notifier available to the background process and dashboard.  
    I can't recommend editing anything in here as you can very easily break something (and it's all internal values so nothing exciting).  
    If you do want to edit, a JSON schema is attached to the file. This means an editor like [VS Code](https://code.visualstudio.com/) can give a nice visualization and completion help.
  
- `.notifier/notifications.json`  
    This file will be created once the first notification is sent.  
    It is a log file that contains the last `n` notifications (`n` is defined by the property `logging.notificationLogSize` in the file `config.js`).  
    Feel free to use the contents of this file for your own projects. You can find the JSON schema in `.vscode/schemas/notifications.json`
  
- `.notifier/cache_manifest.json`  
    This file will be created to maintain cached icons that have been downloaded from an external URL.  
    It really shouldn't be touched but it can safely be deleted to "refresh" icons and fix inconsistencies (TODO: verify that lol).
  
- `.notifier/.env`  
    This file contains your login data and possibly soon other private data. If needed, modify it with the login manager in the control panel.  
    The password is hashed, so even if the file is leaked, nobody can use it to gain access. Still try to adequately protect it though!

<br>

### Advanced configuration:
Node-Notifier contains an even deeper level of configuring; the internal settings file.  
You can find this file at [`src/settings.js`](./src/settings.js) and you can use it to have the finest possible level of control over Node-Notifier.

<br><br>


<br><br><br>


## REST API usage:
Node-Notifier comes with a REST API that is used to trigger desktop notifications.  
This section tells you how you can use it.  
  
If you use an IP address to send these requests to, make sure your router doesn't change the IP over time.  
If the client and server are running on the same device, you can use `127.0.0.1` or `localhost` (or `::1` for IPv6).

<br>

> ### Sending a notification:
> #### `POST /send`
> 
> <br>
> 
> This is the main endpoint, used to send a desktop notification on the device that Node-Notifier is running on.  
> It accepts data in JSON format and has a few query parameters you can set.
> 
> <br><br>
> 
> **Expected request body (JSON):**  
> This is what the request body should look like.  
> Properties inside square brackets are optional.  
> 
> | Property | Description |
> | :-- | :-- |
> | `title` | The (short) title of the notification |
> | `message` | A more detailed message |
> | `[icon]` | A square icon to attach to the notification. Defaults to a placeholder icon on some operating systems if left empty.<br>This image either needs to be present on the system the Node-Notifier server runs on, or you can provide a URL here, which Node-Notifier will automatically download and cache.<br>An `assets` folder will be created at first startup, it is intended for storing both local and cached images. |
> | `[actions]` | Array of actions (buttons or dropdown, depending on OS) the user can select on the notification. Using this automatically enables `?waitForResult` |
> | `[timeout]` | How many seconds to wait before closing the notification automatically. Prioritised over `?waitForResult` - defaults to `10` if left empty. |
> 
> <details><summary><b>Full example (click to view)</b></summary>
> 
> ```json
> {
>     "title": "Door Bell",
>     "message": "Someone is at the door, open it?",
>     "actions": [
>         "Yes",
>         "No"
>     ],
>     "icon": "assets/door.png",
>     "timeout": 42
> }
> ```
> 
> </details>
> 
> <br><br>
> 
> **Query parameters:**
> > `?waitForResult`  
> > This is a value-less parameter that makes Node-Notifier wait until the user reacts on a notification or it times out.  
> > Also, it will make the API return some data on how the user interacted with the notification.  
> > Make sure your HTTP client waits long enough before getting timed out while the API waits for user interaction!  
> >   
> > Example usage:  
> > `POST /send?waitForResult`
> 
> <br><br>
> 
> **Possible response bodies:**
> 
> <details><summary>Base response (click to view)</summary>
> 
> ```json
> {
>     "error": false,
>     "message": "Notification was sent"
> }
> ```
> HTTP status code: `200`
> 
> <br></details>
> 
> 
> <details><summary>Response when using ?waitForResult (click to view)</summary>
> 
> ```json
> {
>     "error": false,
>     "message": "Hello, I am a success message",
>     "result": "Can be a string or null",
>     "type": "Describes the type of notification interaction",
>     "value": "Some other value, idk"
> }
> ```
> HTTP status code: `200`
> 
> <br></details>
> 
> 
> <details><summary>Errored response (click to view)</summary>
> 
> ```json
> {
>     "error": true,
>     "message": "Hello, I am an error message"
> }
> ```
> HTTP status code: `400`
> 
> <br></details>
> 
> 
> <details><summary>Unauthorized response (click to view)</summary><br>
> 
> If the API expects you to authorize with the admin login but you didn't provide it or it is wrong, you will get this response:  
>   
> ```json
> {
>     "error": true,
>     "message": "Node-Notifier requires you to authenticate before accessing this resource"
> }
> ```
> HTTP status code: `401`
> 
> <br></details>
> 
> <br>

<br><br><br>

## Troubleshooting:
You can try these troubleshooting steps if something doesn't work.  
Note that all commands listed here need to be run in the same directory Node-Notifier's package.json file is in.  
  
If this still didn't help you, feel free to [contact me.](https://sv443.net)  

<br>

- **Clear cache files**
    > If you encounter issues using icon URLs (like the image being outdated) you can clear Node-Notifier's cache files:  
    > Within the main directory which contains the package.json file, there should be a folder called `.notifier`  
    > (If you can't see it, your OS has hidden it by default. Please look up how to show hidden files and folders for your operating system and follow those steps.)  
    > This folder may contain the file `cache_manifest.json`, which you can delete to make Node-Notifier re-fetch all icons before they are used the next time.

<br>

- **Reset the `config.yml` file**
    > This config file contains values that you can tweak to configure Node-Notifier to your liking.  
    > If you need to reset it to its default values, just delete it. Even while Node-Notifier is running, it should automatically be regenerated after a second. If not, restart Node-Notifier.  
    > To permanently delete this file, you need to stop or delete the background process and exit the control panel, then try deleting again.

<br>

- **Not autostarting after reboot**
    > This could mean the pm2-installer Windows service or the systemd (or equivalent) hook on other OSes somehow didn't get registered.  
    > To fix this, make sure Node-Notifier is running and then, in a terminal without admin rights, run the commands `pm2 save` and `pm2 startup`  
    > If the `pm2` command isn't recognized, install it with `npm i -g pm2`

<br>

- **Reinstall the background process**
    > If the background process constantly shows its status as `errored` or `stopped` or is otherwise acting up, first try reading the background process log.  
    > To do this, grab the ID or name of the process in the "Manage PM2 process" menu and then run the command `pm2 logs id_or_name_here` in a terminal without admin rights.  
    >   
    > If the logs are inconclusive or you just want to try the good old "did you turn it off and back on again", do this:  
    >   
    > 1. Open Node-Notifier with `npm start`
    > 2. In the control panel, navigate to the "Manage PM2 process" menu
    > 3. Select "Delete process" and confirm with <kbd>Y</kbd>  
    >     Don't be scared, doing this will **not** delete any data, it just uninstalls the background process
    > 4. After it's finished:
    >     - If you're on Windows you need to exit your current terminal and start a new one with admin permissions (<kbd>Windows</kbd> + <kbd>R</kbd>, enter `powershell` and press <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>Enter</kbd>).  
    >         Now enter `npm start` and follow the prompts until the process exits. Now go back to a normal terminal without admin permissions and enter `npm start` again.
    >     - On other operating systems, just enter `npm start` and you should be prompted to install.

<br>

- **Kill the pm2 daemon**
    > If nothing else worked or you are getting the error `connect EPERM //./pipe/rpc.sock`, you can try these steps:  
    >   
    > 1. On Windows, open a terminal with administrator rights (<kbd>Windows</kbd> + <kbd>R</kbd>, enter `powershell` and press <kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>Enter</kbd>).  
    >    On Linux and MacOS sudo or root usually isn't needed for this.
    > 2. Now enter the command `npm run fix`
    > 3. To correctly start up Node-Notifier again **it's important you exit the admin terminal if you're on Windows** and always use a normal terminal from now on, unless you are prompted for it.


<br><br><br>


## Disclaimers:
- The current version of Node-Notifier only supports HTTP.  
    Usually this is fine as long as all devices are part of the same network.  
    But still, a compromised network could allow third parties to read your requests.  
    So make sure you are not on a public network and aren't sending any important data.  
    If I feel like it and there's demand, I might add HTTPS support in the future.

- This software is provided "as is", without any warranty or liability.  
    If you notice something wrong, please [submit an issue](https://github.com/Sv443/Node-Notifier/issues/new/choose) and I'll try my best to take care of it.  
    If you found a security vulnerability, please [contact me directly.](https://sv443.net/)


<br><br>


## Dependencies:
Node-Notifier wouldn't be possible without these awesome projects:
- [axios](https://npmjs.com/package/axios)
- [dotenv](https://npmjs.com/package/dotenv)
- [eslint](https://npmjs.com/package/eslint)
- [express](https://npmjs.com/package/express)
- [fs-extra](https://npmjs.com/package/fs-extra)
- [import-fresh](https://npmjs.com/package/import-fresh)
- [kleur](https://npmjs.com/package/kleur)
- [nanotimer](https://npmjs.com/package/nanotimer)
- [node-notifier](https://npmjs.com/package/node-notifier)
- [node-watch](https://npmjs.com/package/node-watch)
- [open](https://npmjs.com/package/open)
- [pm2](https://npmjs.com/package/pm2)
- [pm2-installer](https://github.com/jessety/pm2-installer)
- [prompts](https://npmjs.com/package/prompts)
- [request-ip](https://npmjs.com/package/request-ip)
- [snyk](https://npmjs.com/package/snyk)
- [svcorelib](https://npmjs.com/package/svcorelib)
- [tcp-port-used](https://npmjs.com/package/tcp-port-used)
- [yaml](https://npmjs.com/package/yaml)


<br><br>


<div style="text-align:center;" align="center">

Made with ❤️ by [Sv443](https://github.com/Sv443)  
Like Node-Notifier? Please consider [supporting me](https://github.com/sponsors/Sv443)

</div>

changelog.md
# Node-Notifier - Changelog

<br>

### Version History:

- **[1.0.0](#100)**

<br>

### Planned Features:
- Enable downloading images from a URL (and cache them locally)
- Add custom pm2 meters so custom values can be displayed in `pm2 desc <id/name>` and `pm2 monit`
- Add video example to the [readme](./README.md#video-showcase)
  
To suggest more, please [open an issue](https://github.com/Sv443/Node-Notifier/issues/new)

<br><br>

---

<br><br>


## 1.0.0
**The initial release**  
  
- Added HTTP backend that listens for these requests:
    - `GET /` to view the dashboard
    - `POST /send` to send a desktop notification
        - `?waitForResult` to make the API wait for an interaction with the notification
        - `?actions` to have a set of actions the user can select
        - Optionally requires basic auth, to not allow unknown clients to trigger notifications
        - Optionally set whitelisted IPs that don't need to provide auth (make sure they are static IPs so no guest device accidentally gets one assigned!)
        - TODO: Automatic downloading and caching of icon URLs
- Added dashboard page TODO: that displays Node-Notifier's status and notifications
    - TODO: verify Dashboard is password protected
- Added pm2 wrapper so the service runs in the background
- Added config file at `./config.js`
- Added update checker TODO: verify works
    - Update instructions in dashboard
    - Desktop notification if update available
- Added menus:
    - Process manager
    - Notification log
    - Login manager


<br><br>


### [EOF]


src/server.js
const express = require("express");
const { resolve } = require("path");
const { unused, allOfType } = require("svcorelib");
const { platform } = require("os");
const { check: portUsed } = require("tcp-port-used");
const { getClientIp } = require("request-ip");

const error = require("./error");
const { hasAuth, respondRequireAuth, isWhitelisted } = require("./auth");
const { getAllProperties, setProperty } = require("./files");
const sendNotification = require("./sendNotification");
const logNotification = require("./logNotification");
const { tryCache } = require("./assetCache");

const { cfg } = require("./config");

/** @typedef {import("node-notifier/notifiers/notificationcenter").Notification} Notification */
/** @typedef {import("express").Request} Request */
/** @typedef {import("express").Response} Response */
/** @typedef {import("./types").JSONCompatible} JSONCompatible */
/** @typedef {import("./types").HttpMethod} HttpMethod */
/** @typedef {import("./types").QueryObj} QueryObj */
/** @typedef {import("./types").RequestBody} RequestBody */


/** Placeholder icon path - relative to project root */
const placeholderIconPath = "./www/favicon.png";

/** URLs that can be accessed with GET */
const getURLs = [ "/", "/int/getproperties" ];
/** URLs that can be accessed with POST */
const postURLs = [ "/send", "/int/setproperty" ];


const app = express();


/**
 * Initializes the express server
 * @returns {Promise<void, string>}
 */
function init()
{
    return new Promise(async (pRes, pRej) => {
        if(await portUsed(cfg.server.port))
            return pRej(new Error(`Error while initializing HTTP server: Port ${cfg.server.port} is already in use. Please kill the process using the port or change the port in 'config.js' to a different number`));

        // serve static files
        app.use(express.static("www"));

        // use JSON middleware
        app.use(express.json());

        app.use((err, req, res, next) => {
            unused(req, next);

            if(typeof err === "string" || err instanceof Error)
            {
                return respondJSON(res, 400, {
                    error: true,
                    message: `General error in HTTP server: ${err.toString()}`
                });
            }
        });

        // CORS & OPTIONS middleware
        app.use((req, res, next) => {
            const allowedMethods = "GET,POST";
            
            // CORS
            res.setHeader("Access-Control-Allow-Origin", "*");
            res.setHeader("Access-Control-Allow-Methods", allowedMethods);

            // OPTIONS
            if(req.method === "OPTIONS")
            {
                res.setHeader("Allow", allowedMethods);
                res.writeHead(200, { "Content-Type": "text/plain; charset=UTF-8" });
                res.end(allowedMethods);

                return;
            }

            next();
        });


        // listen on URLs defined above
        const listener = app.listen(cfg.server.port, () => {
            for(const url of getURLs)
                app.get(url, (req, res) => incomingRequest("GET", req, res, url));
            for(const url of postURLs)
                app.post(url, (req, res) => incomingRequest("POST", req, res, url));

            return pRes();
        });


        // on error
        listener.on("error", err => {
            error("Error while setting up express server", err);
            return pRej(err instanceof Error ? err : new Error(err));
        });
    });
}

/**
 * Called whenever a client's request is received
 * @param {HttpMethod} method
 * @param {Request} req
 * @param {Response} res
 * @param {string} url Lowercase request URL
 */
async function incomingRequest(method, req, res, url)
{
    console.log(`Incoming ${method} request to ${url}`);

    url = url.toLowerCase();

    const clientHasAuth = hasAuth(req);

    // TODO: password protection doesn't work correctly
    switch(method)
    {
    case "GET":
        // dashboard always requires auth
        // TODO: this is served through static middleware so manual auth doesn't work, see https://stackoverflow.com/q/31524804/8602926
        if(url === "/")
        {
            // dashboard always requires auth
            if(clientHasAuth)
                return res.sendFile(resolve("./www/index.html"));
            else
                return respondRequireAuth(res);
        }

        if(clientHasAuth && url === "/int/getproperties")
            return respondJSON(res, 200, (await getAllProperties()));
        else if(!clientHasAuth)
            return respondRequireAuth(res);

        break;

    case "POST":
        return parseRequest(req, res, url);

    default:
        return respondJSON(res, 405, {
            error: true,
            message: `Method '${method}' is not allowed - use GET or POST`
        });
    }
}

/**
 * Sends a JSON object to a client
 * @param {Response} res
 * @param {number} statusCode
 * @param {JSONCompatible} jsonObj
 */
function respondJSON(res, statusCode = 500, jsonObj)
{
    statusCode = parseInt(statusCode);
    if(isNaN(statusCode))
        statusCode = 500;

    if(typeof jsonObj !== "string")
        jsonObj = JSON.stringify(jsonObj, undefined, 4);

    res.writeHead(statusCode, { "Content-Type": "application/json; charset=UTF-8" });
    res.end(jsonObj);
}

/**
 * Called whenever the server recieves a client request with a request body
 * @param {Request} req
 * @param {Response} res
 * @param {string} url
 */
async function parseRequest(req, res, url)
{
    const clientHasAuth = hasAuth(req);
    const whitelisted = isWhitelisted(getClientIp(req));

    const { body } = req;

    /** Is set to true if the client is whitelisted or has provided valid login data or if authentication is disabled in the config */
    const isAuthenticated = whitelisted || cfg.server.requireAuthentication ? clientHasAuth : true;

    if(body === undefined || (body && ((Array.isArray(body) && body.length === 0) || (typeof body === "object" && Object.keys(body).length === 0))))
    {
        return respondJSON(res, 400, {
            error: true,
            message: "No request data was received"
        });
    }

    try
    {
        if(isAuthenticated && url === "/send")
            return sendNotificationRequest(req, res);
        else if(!isAuthenticated)
            return respondRequireAuth(res);

        if(clientHasAuth && url === "/int/setproperty")
        {
            // TODO: add password protection
            try
            {
                const key = body["key"];
                const value = body["value"];

                if(typeof key === "undefined" || typeof value === "undefined")
                {
                    return respondJSON(res, 400, {
                        error: true,
                        message: "Properties 'key' and/or 'value' are missing or invalid"
                    });
                }

                await setProperty(key, value);

                return respondJSON(res, 200, {
                    error: false,
                    message: "Successfully set property"
                });
            }
            catch(err)
            {
                return respondJSON(res, 500, {
                    error: true,
                    message: `Error while setting property: ${err}`
                });
            }
        }
        else if(!clientHasAuth)
            return respondRequireAuth(res);
    }
    catch(err)
    {
        return respondJSON(res, 400, {
            error: true,
            message: `Error while parsing request body: ${err.toString()}`
        });
    }
}

/**
 * Called when a request to '/send' is received
 * @param {Request} req
 * @param {Response} res
 */
async function sendNotificationRequest(req, res)
{
    if(typeof req.body != "object" || Object.keys(req.body).length < 1)
    {
        return respondJSON(res, 400, {
            error: true,
            message: "Request body is empty or invalid"
        });
    }

    /** @type {RequestBody} */
    const body = req.body;

    const { title, message, icon, actions, timeout } = body;

    const invalidProps = [];

    (typeof title !== "string") && invalidProps.push("title");
    (typeof message !== "string") && invalidProps.push("message");
    (typeof icon !== "undefined" && typeof icon != "string") && invalidProps.push("icon");
    (typeof actions !== "undefined" && (!Array.isArray(actions) || !allOfType(actions, "string"))) && invalidProps.push("actions");
    (typeof timeout !== "undefined" && (typeof timeout !== "number" || timeout < 1)) && invalidProps.push("timeout");

    if(invalidProps.length > 0)
    {
        return respondJSON(res, 400, {
            error: true,
            message: `Request body is missing the following ${invalidProps.length === 1 ? "property or it is" : "properties or they are"} invalid: ${invalidProps.join(", ")}`
        });
    }

    let cachedIconPath;

    // download image if it's a URL and not already in the cache
    if(isValidHttpUrl(icon))
        cachedIconPath = (await tryCache(icon))?.path;

    const icn = cachedIconPath ?? icon;

    /** @type {Notification} */
    const iconProps = typeof icon === "string" ? {
        icon: resolve(icn),
        contentImage: resolve(icn),
    } : getDefaultIconProps();


    let timeoutInt = typeof timeout === "number" ? parseInt(timeout) : NaN;

    /** @type {Notification} */
    const timeoutProps = (!isNaN(timeoutInt) && timeoutInt > 0) ? {
        timeout: timeoutInt
    } : {};


    const { waitForResult } = getQueryParams(req); // TODO: not enabled by default when using actions


    /** @type {string[]|undefined} */
    const parsedActions = (Array.isArray(actions) && allOfType(actions, "string")) ? actions : undefined;

    /** @type {Notification} */
    const actionProps = (parsedActions && parsedActions.length > 0) ? { actions: parsedActions, wait: true } : {};


    let resultProps = {};
    let responseMessage = "";


    /** @type {Notification} Notification properties */
    const notifProps = {
        title,
        message,
        ...timeoutProps,
        ...actionProps,
        ...iconProps,
    };

    if(waitForResult) // TODO:FIXME: not working
    {
        try
        {
            const { result, meta } = await sendNotification(notifProps);

            resultProps = {
                result: result || null,
                type: meta.activationType,
                value: meta.activationValue
            };
            responseMessage = "Successfully sent desktop notification and got a result";
        }
        catch(err)
        {
            return respondJSON(res, 400, {
                error: true,
                message: `Error while sending desktop notification: ${err}`
            });
        }
    }
    else
    {
        sendNotification(notifProps).catch(unused);
        responseMessage = "Sent desktop notification";
    }


    // log notification
    try
    {
        if(cfg.logging.logNotifications)
            await logNotification(notifProps);
    }
    catch(err)
    {
        error("Error while logging notification", err);
    }


    return respondJSON(res, 200, {
        error: false,
        message: responseMessage,
        ...resultProps
    });
}

/**
 * Checks if `val` is a valid HTTP URL
 * @param {string} val
 * @returns {boolean}
 */
function isValidHttpUrl(val)
{
    let u;
    try
    {
        u = new URL(val);
    }
    catch (e)
    {
        return false;  
    }
  
    return u.protocol === "http:" || u.protocol === "https:";
}

/**
 * Parses query parameters of a request
 * @param {Request} req
 * @returns {QueryObj}
 */
function getQueryParams(req)
{
    let waitForResult = false;


    if(typeof req.query === "object" && Object.keys(req.query).length > 0)
    {
        // ?waitForResult
        const qWaitForRes = (req.query["waitForResult"]) ? req.query["waitForResult"].toString() : undefined;

        waitForResult = (typeof qWaitForRes === "string" || qWaitForRes === "true" || qWaitForRes === "1" || qWaitForRes === "yes");
    }


    return {
        waitForResult
    };
}

/**
 * Returns the default icon properties of a notification.  
 * Does checks for OS, because some notification binaries come with their own placeholders.
 * @returns {Partial<Notification>}
 */
function getDefaultIconProps()
{
    if(!cfg.notifications.placeholderIconEnabled)
        return {};

    const plat = platform();

    let needsPlaceholder = true;

    if(plat === "darwin") // MacOS
        needsPlaceholder = false;

    return needsPlaceholder ? {
        icon: resolve(placeholderIconPath),
        contentImage: resolve(placeholderIconPath),
    } : {};
}

module.exports = { init };
